File: coding/src/com/algos/core/li13_graph/revised/ShortestPath_Dijkstras.java
package  com.algos.core.li13_graph.revised;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.Queue;

public class ShortestPath_Dijkstras {

	public static void main(String[] args) {
		ShortestPath_Dijkstras shortedpath = new ShortestPath_Dijkstras();
		
		int[][][] adj = new int[][][]{{{1, 1}, {2, 6}}, {{2, 3}, {0, 1}}, {{1, 3}, {0, 6}}};
            
		int dist[] = shortedpath.dijkstra(adj, 2);
		System.out.println(Arrays.toString(dist));
	}

	// Idea: Same as prims algo except that instead of just check the edge weight we track cumulative sum from root
	// Use dist array to track distance each vertices
	// May be a parent array if you want to print path
	// Drawbacks: Won't work with negative edge cycles. Use Bellman Ford algo.
	//O(ElogV)
	public int[] dijkstra(int[][][] adj, int src) {
		int n = adj.length;
		
		Queue<int[]> pq = new PriorityQueue<>((d1, d2)->d1[0]-d2[0]);
		int[] dist = new int[n];
		Arrays.fill(dist, Integer.MAX_VALUE);

		dist[src] = 0;
		pq.add(new int[]{0, src});	//{dist, node}
		while(!pq.isEmpty()){
			int[] curr = pq.remove();
			int dis = curr[0], node = curr[1];
			
			for (int nei[]: adj[node]) {
				int neiNode = nei[0], edgeWeight = nei[1];
				if (dis + edgeWeight < dist[neiNode]) {	// visited is not needed as anyway this condition won't satisfy for already visited nodes
					dist[neiNode] = dis + edgeWeight;
					pq.add(new int[]{dist[neiNode], neiNode});
				}
			}
		}
		return dist;
	}

	// print parent
	// node = dest
	// while(parent[node] != node)
	// 		node = parent[node]
}

File: dsa/patterns/13_graphs/python/shortest_path_dijkstras.py
import heapq
from typing import List


"""
Idea: Same as Prim's algorithm except that instead of just checking the edge
weight, we track cumulative distance from the source.
Use dist array to track distance of each vertex.
Parent array can be used if path reconstruction is needed.
Drawbacks: Won't work with negative edge weights. Use Bellman-Ford instead.
Time Complexity: O(E log V)
"""
def dijkstra(adj: List[List[List[int]]], src: int) -> List[int]:
    n = len(adj)
    INF = float("inf")

    dist = [INF] * n
    dist[src] = 0

    pq = []
    heapq.heappush(pq, (0, src))   # (distance, node)

    while pq:
        curr_dist, node = heapq.heappop(pq)

        for nei_node, weight in adj[node]:
            if curr_dist + weight < dist[nei_node]:
                dist[nei_node] = curr_dist + weight
                heapq.heappush(pq, (dist[nei_node], nei_node))

    return dist


# one test (same as Java main)
if __name__ == "__main__":
    adj = [
        [[1, 1], [2, 6]],
        [[2, 3], [0, 1]],
        [[1, 3], [0, 6]]
    ]

    print(dijkstra(adj, 2))
